### ISSUE TITLE
Changelog/Contributing discussion

### ISSUE THREAD

**Issue by m0ngr31** [m0ngr31/kodi-alexa#63]
I wanted to open this up and get opinions on what the policies for these are. I originally wanted the changelog to discuss new features. I don't see an issue with using something like semver for as many small features that are added. This gets a little tricky when there are 4 or 5 PRs open at once, so maybe there should be an version number increase when a few things have been merged at a time.

I didn't want it to become a alternative for `git log`. I was just hoping for something simple for people who maybe aren't developers or don't use git on a regular basis to see the current features vs what they have now.

We should probably start using tags on new releases as well so it's easier to help see where someone is at if they have a problem.

**Comment by jingai** [comment]
I have been adding tags since I saw they were missing. Github automatically zips these up as releases too, which I think would be good to point most users at so they don't have to mess with Git at all if they don't want to. 

I agree that the changelog shouldn't be a copy of or replacement for the Git history. IMHO it should be a curated list of things we think end-users will find interesting.

**Comment by digiltd** [comment]
How about all PR changes get added under the title "unreleased"? Or not make it a contribution requirement and rely on Git history. Then when it is decided that a new release version is needed, you can curate the list of features from the unreleased list of changes.

Some changes would need to be made to the tutorial and forum post to make clear what a user is supposed to download (the readme currently tells the user to clone the master branch). Adding release tags work if the user knows what they are, if they don't then they would automatically click the biggest download button which will download the master branch. 

What are your thoughts on using a dev branch for work in progress and use master as the stable release? This would simplify things for a non developer/git user because either they download something from the releases tab, clone the repo or download it via the big green "clone or download" button. But they would all basically be getting the same version unless they switched branches.

**Comment by jingai** [comment]
> Or not make it a contribution requirement and rely on Git history. Then when it is decided that a new release version is needed, you can curate the list of features from the unreleased list of changes.

I don't personally mind doing this and handling the release tags and changelog.

> Some changes would need to be made to the tutorial and forum post to make clear what a user is supposed to download (the readme currently tells the user to clone the master branch). Adding release tags work if the user knows what they are, if they don't then they would automatically click the biggest download button which will download the master branch.

If we go this route, I'd modify the README to direct the users at the Releases tab at the top and to download the first ZIP.  For that matter, we could just link directly to the ZIP archive in the README and keep it up-to-date.  I don't mind maintaining this either.

> What are your thoughts on using a dev branch for work in progress and use master as the stable release?

I don't personally like this because it differs from most other projects out there.  There is value in following common practices, particularly for attracting new developers.

IMHO, the README just needs to not direct people at git at all.  Those that know they want the unstable releases can likely figure out what to do.

**Comment by digiltd** [comment]
> I don't personally like this because it differs from most other projects out there. There is value in following common practices, particularly for attracting new developers.

I don't mind either way, but I find it surprising you say it is an uncommon practice, it seems to be used by loads of large projects (XBMC, SiLVO, OpenHAB, Gulp, npm, home-assistant, Kirby, Bootstrap, pi-hole, ha-bridge, Foundation, Angular, Docker, Sass). 

If you are happy to maintain it then maybe leave things as they are. People contribute and you can continue to give your helpful (and genuinely appreciated) guidance where needed so contributed changes "fit".

> IMHO, the README just needs to not direct people at git at all. Those that know they want the unstable releases can likely figure out what to do.

Is it possible to set everything up just using the zip? I think the need to use Python, pip, awscli, lambda-deploy means that this project does require a certain level of knowledge from the start.  

Whilst it is not directly related to this issue, it would be good to bring the topic of documentation to the table.

Due to the desirable end result that this Skill delivers it does attract a lot of users who are not developers, though I suppose Kodi users are generally more technically minded than most. As the project grows I think we do need to work on the documentation (I am more than happy to contribute) to try to remove any assumptions about what the user already knows and give anyone reading enough information to take things further (even if just a nod in the right direction).

It will be very valuable to those that need it, and those that don't can just skim through.

One example of this is how we tell users they to open up their Kodi box to the internet via port forwarding using unsecure standard http without touching on the security side of things. Granted we don't need to delve deep into network security, but we should at least inform the user that the opening up of ports to the internet is a pretty big security risk and to proceed with caution.

Another would be about self hosting. I am more than comfortable setting up my own server, in  the 20 years I have been working on the web, I have setup hundreds of sites on various platforms in languages that have come and gone (php, html, xhtml, shtml (retro), Perl (even more retro), asp (shudder)). Yet, I have never come across wsgi or how to set it up until I started using this skill. So whilst I was fine setting up my own Apache/Nginx server, it was all new ground (and a bit of a faff) working out the wsgi bit :) 

Lots of this stuff could be referenced in the wiki if you didn't want to have to scroll through a super long readme.

**Comment by jingai** [comment]
> > I don't personally like this because it differs from most other projects out there. There is value in following common practices, particularly for attracting new developers.
> 
> I don't mind either way, but I find it surprising you say it is an uncommon practice, it seems to be used by loads of large projects (XBMC, SiLVO, OpenHAB, Gulp, npm, home-assistant, Kirby, Bootstrap, pi-hole, ha-bridge, Foundation, Angular, Docker, Sass).

Kodi (XBMC) definitely doesn't do it this way.  master is their development (mainline) branch.  They make branches for _stable_ releases.  I'm proposing we just make use of tags for that, since they are actually branches underneath anyway AFAIR, for the purposes of checkouts at least.  If we get to the point that we want to also maintain stable releases, we'd need branches we can continue to push to.. but I don't think that's necessary yet.

A couple others you referenced don't seem to do it the way you're suggesting, either.  And https://github.com/home-assistant/home-assistant it seems actually just renamed master to dev for some weird reason.  Loading their repo without a branch specified gets you one called 'dev', which means the 'clone or download' button would still point to mainline anyway..

I realize it _can_ be done this way, but the best argument for _not_ doing it like this is simply that I know where to start hacking when I enter a project if everyone uses master for the same purpose.  Having to hunt through docs/wiki/forums just to discover the mainline branch is annoying.

Your argument seems to stem from the fact that the 'default' branch the users see is the mainline, which encourages them to pull that.  But here's the thing:  github is _not_ an end-user site; it's for developers.  We shouldn't be releasing through git at all if possible.

> Is it possible to set everything up just using the zip? I think the need to use Python, pip, awscli, lambda-deploy means that this project does require a certain level of knowledge from the start.

It does, but we can remove git from the equation at least.  This skill, in its current form at least, will never be easy for the average user to set up.

> Due to the desirable end result that this Skill delivers it does attract a lot of users who are not developers, though I suppose Kodi users are generally more technically minded than most. As the project grows I think we do need to work on the documentation (I am more than happy to contribute) to try to remove any assumptions about what the user already knows and give anyone reading enough information to take things further (even if just a nod in the right direction).

Agreed, but we also need to discuss scope.  Given we're essentially telling users to set up a web server, that rabbit hole is quite deep (and potentially dangerous).

> One example of this is how we tell users they to open up their Kodi box to the internet via port forwarding using unsecure standard http without touching on the security side of things. Granted we don't need to delve deep into network security, but we should at least inform the user that the opening up of ports to the internet is a pretty big security risk and to proceed with caution.

The flip-side of this is the nightmare of dealing with helping people to secure it.  Remember, we don't have a support hotline staffed by a bunch of people taking trouble tickets.  It's just a couple of us on forums.

This whole thing is dangerous without solid prior knowledge of a decent number of different topics.  Unless Amazon will accept this skill in some form though, I am not sure how to get around that.

...a big help would be if Kodi's web server simply supported HTTPS.. I'm honestly not sure how it doesn't at this point.

> Another would be about self hosting.

It's my opinion that we probably should just avoid mentioning this at all.  Running your own internet-facing web server is a big responsibility, and I think we're taking it too lightly.

Perhaps a warning would suffice, but I think it should be a sufficiently visible warning..

**Comment by digiltd** [comment]
I agree. Sorry, I know the list of repos does appear a little petty, and I debated sending it, but when you mentioned it wasn't standard I was curious to see and started to see what some of other projects I use do. Generally I work alone and like many developers I learn from what I see others doing and it is all in the interest of healthy discussion :)

To clarify, I don't like (and am not suggesting) having a bunch of stagnant branches with one for each release (as some of the examples do) because that is what tags and releases are for. But there is value in using a dev branch.

And yes, we don't want to have to support all the various areas that this project touches, and a warning/disclaimer would help alot. As I don't see how we can avoid covering at least the basic risks plus pros and cons of the various hosting methods.

I have a good solution to the security issue, using Nginx as a proxy for signed ssl traffic into an unconventional port on a private ddns server, allowing only connections from Amazon IPs from a specific region (i'm jonjon on the forum). Which I am happy to share, I have also read about others using MQTT to avoid opening ports at all. I haven't looked into it but it might be easier to provide setup instructions for, and it gives some control by letting them disconnect from the broker.

What are the traffic stats for the repo like?  Interest on the forum has slowed down, maybe it generally not ready for prime time. But it would be cool if at some point we could offer a way to make it secure and accessible to all.

Food for thought.

**Comment by jingai** [comment]
> To clarify, I don't like (and am not suggesting) having a bunch of stagnant branches with one for each release (as some of the examples do) because that is what tags and releases are for.

Actually, this is useful if you plan to maintain multiple versions.  For instance, with skin.metropolis, I tend to maintain versions for the most recent version of Kodi and the previous version.  So I do keep branches around and name them to match the Kodi releases (Jarvis, Krypton, etc).  For some issues pertaining even older releases, I'll sometimes hit them if the fix is a) simple, and b) fixes something that broke it badly (crashes on startup, won't display menus, etc).

For this particular project though, I don't think we need to do that -- at least not yet.  Most issues that crop up are pretty isolated, so just fixing and releasing another version works fine.

> But there is value in using a dev branch.

Yes, for very large projects with regular (longish) release schedules, I can see the use.  But that's not us, so I feel there's more value in using the default branch (that is, the one created by git when initializing the repo) as the mainline branch, and hanging all of the feature/fix/release branches off that.

Honestly, this is a pretty hotly-debated topic and everyone has their own opinions.  From my perspective, what has mattered most to me is to be able to know at a glance what branch to start hacking on.  Often I'll poke around at a project before knowing for certain that I want to become a contributor.  Usually I've encountered a bug and want to have a go at fixing it.  If I have to read through their contributing document or worse before I can even poke around, I'll usually just look to the next project if one is available.

> And yes, we don't want to have to support all the various areas that this project touches, and a warning/disclaimer would help alot. As I don't see how we can avoid covering at least the basic risks plus pros and cons of the various hosting methods.

Yeah, the problem here is scope.  I'm in agreement that we should avoid duplicating proper documentation -- the devil is in the details for a lot of the stuff that can bite you -- for the various bits we hook into, and instead provide sufficient warnings in the README.  Links to some good documentation would be welcome, but I _really_ don't think someone should be setting up something like a web server with only a README from an Alexa skill.

For Lambda/Heroku (i.e., cloud) deployments, this particular part is less of a risk to the end-user, but I'm always shocked at how tangled up some users can get..

The Kodi endpoint remains an issue, and...

> I have a good solution to the security issue, using Nginx as a proxy for signed ssl traffic into an unconventional port on a private ddns server, allowing only connections from Amazon IPs from a specific region (i'm jonjon on the forum). Which I am happy to share, I have also read about others using MQTT to avoid opening ports at all. I haven't looked into it but it might be easier to provide setup instructions for, and it gives some control by letting them disconnect from the broker.

...stuff like this definitely helps, but there are still situations where the user needs to have some idea of what it is they're actually doing.  For one, simple typos when following your guide can cause them grief.  For another, updating the web server.. and updating the configuration if any of the directives changed as a result of the update.. or if defaults changed, etc..

But again, I think all we can do is provide a warning.

> What are the traffic stats for the repo like? Interest on the forum has slowed down, maybe it generally not ready for prime time. But it would be cool if at some point we could offer a way to make it secure and accessible to all.

I'm not sure how/if we can see people that just download the code.  Forks we can see.

But I think interest will taper off simply due to the difficulty of setup anyway.  There's a lot of ways things can go wrong, no matter how much we spell out in the docs.

A discussion about how to simplify the set up would be good, but maybe out of scope for this particular Issue.

**Comment by jingai** [comment]
Maybe it would help to actually write down what I'm accustomed to as a developer..

I tend to think of the master branch as mostly stable (a la, Debian's 'unstable'), in that each commit was tested in isolation, but perhaps not thoroughly tested in concert with the rest of the project.

Tagged releases (or branches) are 'stable.'  These are things I believe (or believed at the time of release) to be working 100%.  These are snapshots of master at a point in time where everything was working and contained enough new features to deserve a version bump.

Development branches are our individual branches we create on our own forks.

For collaborative development on a given feature, I'd generally make a new branch on upstream with a name describing the feature/bugfix or add someone as a collaborator to my own fork and work there.

If I just want to show someone my work, I'll push to a branch on origin (my own fork) and link them to that.

For a submission, that's obviously what PRs are for.

After writing all of this down, I guess what I dislike most about a 'develop' branch is that I don't think there should be only one.  And since we all as developers already have our own forks/branches for development, why do we need one upstream?

Until/if we add branches for stable releases, tags serve as our 'stable' and the master branch serves as our mainline development branch.  There is no need for a third until we decide we need to _maintain_ stable releases, and even then, why not just push to the stable branch(es) we'd already have anyway?

You could, of course, do as I described in the previous paragraph but with stable and unstable reversed.. but that in and of itself is just a rename, and doing it the way I described means the workflow remains the same with or without the concept of a maintained stable branch.

**Comment by digiltd** [comment]
> but I really don't think someone should be setting up something like a web server with only a README from an Alexa skill.

Probably true, but I think the reward (especially seeing it in action in the videos) is enough for many to want to have a go at doing what it takes to get it working.

The Nginx setup I mentioned is to securely receive the requests from Lambda to my Kodi. It was a bit of a faff but I didn't want to set up the Skill until I knew it was solid (I did a quick unsecure test to see if it was worth it first though).

> I'm not sure how/if we can see people that just download the code. Forks we can see.

It should be in the Graphs tab, and then under Traffic. 

https://help.github.com/articles/about-repository-graphs/#traffic

Cheers for describing your workflow. If it helps my typical workflow for developing sites is to have the master branch, that is the live site and I deploy to the webserver from that branch. Then I have the staging branch, which deploys to.... you guessed it, the staging server. I create a branch from staging when working on a feature/change/fix, then merge to staging (deleting the feature branch) and when all is tried, tested and signed off, I merge staging into master.

This discussion started when talking about a stable(ish) release but still having a bleeding edge version we can suggest a user tries while something is being changed or added. I was seeing the master branch as being the stable(ish) "live" version with another branch used in the same way as my "staging" branch. For those things that whilst tested on our local environments, still need a little longer to iron the kinks out, a communal "develop" branch if you like. PR are made against the dev (or whatever it is called) branch.

Now that I understand your approach I agree it doesn't offer a significant advantage over using the master branch in the same communal way. It does (imho) make it a little easier to visually (and mentally) separate the two, especially for someone who's Git level doesn't really go much above making a branch, merging a branch. 

With that said, I'm happy park the branching issue :)

**Comment by jingai** [comment]
Yeah, at least so far, we don't have the concept of a testing/live branch here (at least, not one that's separate from the master branch).  It might be a good idea to have one, but it's my preference to get the testing out of the way in the PR instead; i.e., before it's merged to master.

No one is _ever_ certain that things are 100%, but I like being as proactive about it as possible.

What this means is, master is actually generally OK to use most of the time.  For users that have a hard time debugging things though, maybe they'd want to stay on a release that's seen more use out in the wild first.  For them, tagged releases are the way to go.

Some users don't even know enough to provide the necessary information to us to debug their problem.  That's really the situation I want to (try to) prevent, as it translates to a _lot_ of time expended for even simple issues.

As for traffic, it's 121 clones and 488 unique visitors in the last 14 days.  I don't see a way to look further back.

**Comment by jingai** [comment]
> Cheers for describing your workflow. If it helps my typical workflow for developing sites is to have the master branch, that is the live site and I deploy to the webserver from that branch. Then I have the staging branch, which deploys to.... you guessed it, the staging server. I create a branch from staging when working on a feature/change/fix, then merge to staging (deleting the feature branch) and when all is tried, tested and signed off, I merge staging into master.

One final comment and then I'll shut up about it:  this is essentially what I do, only with staging and master reversed.  My logic for doing it this way is that it works (as in, I maintain my workflow) with the presence of any combination of branches (staging/testing/live, stable, etc), since master is, well, the master.. it has everything.

Wanted to also thank you for the civil discussion on this.  It's a topic I hadn't really thought much about in a while, so it was useful.. but so often these sorts of 'religious' topics turn ugly quickly :)

**Comment by digiltd** [comment]
Likewise :)

**Comment by m0ngr31** [comment]
At my job, we have the `master` branch as the always stable branch. We use develop for all development/testing. All new features and bugfixes are on feature branches based on master.

For this project though, I'm fine basing everything on the master branch so that the bulk of people using this that aren't developers aren't waiting on features and fixes that toil away on the develop branch while we prepare for a release.

**Comment by jingai** [comment]
> All new features and bugfixes are on feature branches based on master.

That's kind of the crux of it though.  If you use develop for all development, then.. how are you also using branches based on master?  I understand having a testing branch (it's useful for pulling in all of the other branches as a whole), but not sure that for this we need a communal one.  I guess if we get a few more developers, maybe.

Basically, I think what you mean is master == stable but you also have a testing branch at work?

As far as master being stable, it tends to be that anyway provided we review PRs well enough.  The biggest problem I'm personally encountering is that I don't have it deployed via Lambda/Heroku/etc (and really don't want to bother), so, while I do try to wait for someone to test, it doesn't always happen until after it's "out in the wild."  The option here is to hold the feature potentially indefinitely on a branch, or push it into master and see what happens.  The latter means that master is "stable-but-maybe-not-for-all-configurations."

Either way, I guess what we really need to decide is how often we want to tag releases.  If it's very frequent, we don't need to change much.  If we version less often (which is my preference), I'd suggest we at least make mention in the README that the tagged releases can be "more stable" for those that prefer a problem-free experience over having all of the latest features.

**Comment by m0ngr31** [comment]
> That's kind of the crux of it though. If you use develop for all development, then.. how are you also using branches based on master?

Sorry, that was a typo on my part. We use develop as the branch that all work is done off of. Master only gets updated when develop is merged in (or a hotfix).

> Either way, I guess what we really need to decide is how often we want to tag releases. If it's very frequent, we don't need to change much. If we version less often (which is my preference), I'd suggest we at least make mention in the README that the tagged releases can be "more stable" for those that prefer a problem-free experience over having all of the latest features.

I'm fine just tagging every once in a while, but I would like a changelog/readme update when something is tagged. Does that work?

**Comment by jingai** [comment]
> I'm fine just tagging every once in a while, but I would like a changelog/readme update when something is tagged. Does that work?

It's your repo -- I was just stating my opinions on things :)

But yes, that works for me, as that's what I'd planned on doing anyway if we'd decided to defer releases (batch up features/fixes).

I don't mind handling this, unless you want to.  Shared responsibility is fine too.  Just want to make sure you're OK with me deciding when to release.  If not, I can make PRs for that when I think one is due, or I can just leave it up to you.  Your call.

**Comment by m0ngr31** [comment]
> It's your repo -- I was just stating my opinions on things :)

Yes, but I have really appreciated your contributions both in code and in the community. This skill wouldn't be where it is without your help. So your input is very important to me.

I'm fine keeping things going the way they are for right now as long as the documentation isn't falling behind. I could be wrong, but it almost feels like a lot of intents are almost hidden because there isn't any documentation about them outside of the utterances file. Maybe I need to create a tutorial file about the different intents capabilities and how to invoke them.

**Comment by jingai** [comment]
Yeah, documentation is falling behind.  Mostly because with the limited time I have to work on stuff like this, it's obviously more gratifying to create new functionality.

But, cleaning up both code and documentation is on the list of things I want to get to.  Just had a few things I was trying to sort through first.

**Comment by m0ngr31** [comment]
I'm right there with you. You can see that I'm pretty sporadic when it comes to replying to all of these issues, let alone updating docs. I just don't want to see them get left behind.

**Comment by m0ngr31** [comment]
I'll be changing this in the 2.5 update.
